File: app/src/main/java/it/ibashkimi/lockscheduler/settings/SettingsFragment.java
package it.ibashkimi.lockscheduler.settings;

import android.content.Intent;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.XmlRes;
import android.support.v4.app.DialogFragment;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceCategory;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.widget.GridLayoutManager;
import android.text.InputType;
import android.view.View;

import com.afollestad.materialdialogs.MaterialDialog;

import java.util.List;

import it.ibashkimi.lockscheduler.R;
import it.ibashkimi.lockscheduler.util.Utils;
import com.ibashkimi.support.preference.ThemeAdapter;
import com.ibashkimi.support.preference.ThemeItem;
import com.ibashkimi.support.preference.ThemePreference;
import com.ibashkimi.support.preference.ThemePreferenceDialogFragmentCompat;
import com.ibashkimi.support.preference.Themes;


public class SettingsFragment extends PreferenceFragmentCompat {
    private static final int REQUEST_CODE_ALERT_RINGTONE = 0;

    private MaterialDialog themeDialog;

    @Override
    public void onCreatePreferences(Bundle bundle, String s) {
        getPreferenceManager().setSharedPreferencesName("settings");
        addPreferencesFromResource(R.xml.preferences);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void addPreferencesFromResource(@XmlRes int preferencesResId) {
        super.addPreferencesFromResource(preferencesResId);
        // Colored navigation bar
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP && !Utils.hasNavBar(getContext())) {
            PreferenceCategory category = (PreferenceCategory) findPreference("appearance");
            category.removePreference(findPreference("colored_navigation_bar"));
        }

        findPreference("min_pin_length").setSummary("" + getIntPreference("min_pin_length", 4));
        findPreference("min_password_length").setSummary("" + getIntPreference("min_password_length", 4));
    }

    @Override
    public void onDisplayPreferenceDialog(Preference preference) {
        DialogFragment dialogFragment = null;
        if (preference instanceof ThemePreference) {
            dialogFragment = ThemePreferenceDialogFragmentCompat.newInstance(preference.getKey());
        }

        // If it was one of our custom Preferences, show its dialog
        if (dialogFragment != null) {
            dialogFragment.setTargetFragment(this, 0);
            dialogFragment.show(getFragmentManager(),
                    "android.support.v7.preference" + ".PreferenceFragment.DIALOG");
        } else {
            super.onDisplayPreferenceDialog(preference);
        }
    }

    @Override
    public boolean onPreferenceTreeClick(Preference preference) {
        switch (preference.getKey()) {
            case "notifications_ringtone":
                Intent intent = new Intent(RingtoneManager.ACTION_RINGTONE_PICKER);
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_NOTIFICATION);
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT, true);
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT, true);
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI, Settings.System.DEFAULT_NOTIFICATION_URI);
                String existingValue = getRingtonePreferenceValue();
                if (existingValue != null) {
                    if (existingValue.length() == 0) {
                        // Select "Silent"
                        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, (Uri) null);
                    } else {
                        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, Uri.parse(existingValue));
                    }
                } else {
                    // No ringtone has been selected, set to the default
                    intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, Settings.System.DEFAULT_NOTIFICATION_URI);
                }
                startActivityForResult(intent, REQUEST_CODE_ALERT_RINGTONE);
                return true;
            case "theme":
                final List<ThemeItem> themes = Themes.getThemeItems();
                int savedThemeId = getPreferenceManager().getSharedPreferences().getInt("theme", Themes.Theme.APP_THEME_DAYNIGHT_INDIGO);
                int themeIndex = -1;
                for (int i = 0; i < themes.size(); i++) {
                    if (themes.get(i).id == savedThemeId) {
                        themeIndex = i;
                        break;
                    }
                }
                ThemeAdapter themeAdapter = new ThemeAdapter(getContext(), Themes.getThemeItems(), themeIndex, new ThemeAdapter.ThemeSelectedListener() {
                    @Override
                    public void onThemeSelected(ThemeItem item) {
                        getPreferenceManager().getSharedPreferences().edit().putInt("theme", item.id).apply();
                        if (themeDialog != null) themeDialog.dismiss();
                        themeDialog = null;
                    }
                });
                themeDialog = new MaterialDialog.Builder(getContext())
                        .title(R.string.pref_title_theme)
                        // second parameter is an optional layout manager. Must be a LinearLayoutManager or GridLayoutManager.
                        .adapter(themeAdapter, new GridLayoutManager(getContext(), 2))
                        .negativeText(R.string.dialog_action_cancel)
                        .itemsCallbackSingleChoice(themeIndex, new MaterialDialog.ListCallbackSingleChoice() {
                            @Override
                            public boolean onSelection(MaterialDialog dialog, View view, int which, CharSequence text) {
                                setPreference("theme", themes.get(which).id);
                                return true;
                            }
                        })
                        .show();
                return true;
            case "min_password_length":
                new MaterialDialog.Builder(getContext())
                        .title(R.string.pref_min_password_length_dialog_title)
                        .inputType(InputType.TYPE_CLASS_NUMBER)
                        .input("", "" + getIntPreference("min_password_length", 4), new MaterialDialog.InputCallback() {
                            @Override
                            public void onInput(@NonNull MaterialDialog dialog, CharSequence input) {
                                setPreference("min_password_length", Integer.parseInt(input.toString()));
                            }
                        }).show();
                return true;
            case "min_pin_length":
                new MaterialDialog.Builder(getContext())
                        .title(R.string.pref_min_pin_length_dialog_title)
                        .inputType(InputType.TYPE_CLASS_NUMBER)
                        .input("", "" + getIntPreference("min_pin_length", 4), new MaterialDialog.InputCallback() {
                            @Override
                            public void onInput(@NonNull MaterialDialog dialog, CharSequence input) {
                                setPreference("min_pin_length", Integer.parseInt(input.toString()));
                            }
                        }).show();
                return true;
        }
        return super.onPreferenceTreeClick(preference);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE_ALERT_RINGTONE && data != null) {
            Uri ringtone = data.getParcelableExtra(RingtoneManager.EXTRA_RINGTONE_PICKED_URI);
            if (ringtone != null) {
                setRingtonePreferenceValue(ringtone.toString());
            } else {
                // "Silent" was selected
                setRingtonePreferenceValue("");
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Nullable
    private String getRingtonePreferenceValue() {
        return getStringPreference("notifications_ringtone", null);
    }

    private void setRingtonePreferenceValue(String ringtone) {
        setPreference("notifications_ringtone", ringtone, false);
    }

    /*private void setPreference(String key, String value) {
        setPreference(key, value, true);
    }*/

    private void setPreference(String key, int value) {
        setPreference(key, value, true);
    }

    private void setPreference(String key, String value, boolean updateSummary) {
        getPreferenceManager().getSharedPreferences().edit().putString(key, value).apply();
        if (updateSummary)
            findPreference(key).setSummary(value);
    }

    private void setPreference(String key, int value, boolean updateSummary) {
        getPreferenceManager().getSharedPreferences().edit().putInt(key, value).apply();
        if (updateSummary)
            findPreference(key).setSummary(Integer.toString(value));
    }

    private String getStringPreference(String key, String defaultValue) {
        return getPreferenceManager().getSharedPreferences().getString(key, defaultValue);
    }

    private int getIntPreference(String key, int defaultValue) {
        return getPreferenceManager().getSharedPreferences().getInt(key, defaultValue);
    }
}

File: app/src/main/java/com/ibashkimi/lockscheduler/settings/SettingsFragment.kt
package com.ibashkimi.lockscheduler.settings

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.text.InputType
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.Toast
import androidx.annotation.XmlRes
import androidx.navigation.Navigation.findNavController
import androidx.navigation.ui.NavigationUI
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import androidx.preference.PreferenceFragmentCompat
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.input
import com.ibashkimi.lockscheduler.R
import com.ibashkimi.lockscheduler.model.action.LockAction
import com.ibashkimi.lockscheduler.data.prefs.AppPreferencesHelper
import com.ibashkimi.lockscheduler.ui.BaseActivity
import com.ibashkimi.lockscheduler.util.*
import com.ibashkimi.theme.activity.ThemePreferences
import com.ibashkimi.theme.theme.NavBarColor
import com.ibashkimi.theme.theme.NightMode


class SettingsFragment : PreferenceFragmentCompat(),
    SharedPreferences.OnSharedPreferenceChangeListener {

    private val sharedPreferences: SharedPreferences by lazy {
        context!!.getSharedPreferences(AppPreferencesHelper.PREFERENCES_NAME, Context.MODE_PRIVATE)
    }

    private var lockTypeIfGranted: LockAction.LockType
        get() {
            val rep =
                sharedPreferences.getString("lock_if_granted", LockAction.LockType.UNCHANGED.name)!!
            return LockAction.LockType.values().first { it.value == rep }
        }
        set(value) = sharedPreferences.edit().putString("lock_if_granted", value.name).apply()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.settings, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val navController = findNavController(requireActivity(), R.id.main_nav_host_fragment)
        return NavigationUI.onNavDestinationSelected(item, navController)
                || super.onOptionsItemSelected(item)
    }

    override fun onStart() {
        super.onStart()
        AppPreferencesHelper.preferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onStop() {
        super.onStop()
        AppPreferencesHelper.preferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        val themeActivity = requireActivity() as BaseActivity
        when (key) {
            ThemePreferences.KEY_NIGHT_MODE -> themeActivity.applyNightMode(
                themeActivity.themePreferences
                    .getNightMode(NightMode.DAYNIGHT)
            )
            ThemePreferences.KEY_NAV_BAR_COLOR -> themeActivity.applyNavBarColor(
                themeActivity.themePreferences
                    .getNavBarColor(NavBarColor.SYSTEM)
            )
            "loitering_delay" -> Toast.makeText(
                requireContext(),
                "Not implemented yet",
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    override fun onCreatePreferences(bundle: Bundle?, s: String?) {
        preferenceManager.sharedPreferencesName = AppPreferencesHelper.PREFERENCES_NAME
        addPreferencesFromResource(R.xml.preferences)
    }

    override fun addPreferencesFromResource(@XmlRes preferencesResId: Int) {
        super.addPreferencesFromResource(preferencesResId)
        // Colored navigation bar
        if (!Utils.hasNavBar(requireContext())) {
            findPreference<PreferenceCategory>("appearance")
                ?.removePreference(findPreference("colored_navigation_bar"))
        }

        findPreference<Preference>("min_pin_length")?.summary =
            "" + getIntPreference("min_pin_length", 4)
        findPreference<Preference>("min_password_length")?.summary =
            "" + getIntPreference("min_password_length", 4)
        val rep = getStringPreference("lock_at_boot", LockAction.LockType.UNCHANGED.name)
        updateSummary(LockAction.LockType.valueOf(rep))
    }

    override fun onPreferenceTreeClick(preference: Preference): Boolean {
        when (preference.key) {
            "notifications" -> {
                val intent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    Intent(Settings.ACTION_CHANNEL_NOTIFICATION_SETTINGS).apply {
                        putExtra(Settings.EXTRA_APP_PACKAGE, requireContext().packageName)
                        putExtra(
                            Settings.EXTRA_CHANNEL_ID,
                            PROFILE_ACTIVATED_NOTIFICATION_CHANNEL_ID
                        )
                    }
                } else {
                    val intent = Intent()
                    intent.action = "android.settings.APP_NOTIFICATION_SETTINGS"

                    //for Android 5-7
                    intent.putExtra("app_package", requireContext().packageName)
                    intent.putExtra("app_uid", requireContext().applicationInfo.uid)

                    // for Android 8 and above
                    //intent.putExtra("android.provider.extra.APP_PACKAGE", getPackageName())
                }
                startActivity(intent)
                return true
            }
            "theme" -> {
                findNavController(requireActivity(), R.id.main_nav_host_fragment)
                    .navigate(R.id.action_settings_to_themeFragment)
                return true
            }
            "min_password_length" -> {
                MaterialDialog(requireContext()).show {
                    title(R.string.pref_min_password_length_dialog_title)
                    input(
                        inputType = InputType.TYPE_CLASS_NUMBER,
                        prefill = getIntPreference("min_password_length", 4).toString()
                    ) { _, text ->
                        setPreference("min_password_length", Integer.parseInt(text.toString()))
                    }
                }
                return true
            }
            "min_pin_length" -> {
                MaterialDialog(requireContext()).show {
                    title(R.string.pref_min_pin_length_dialog_title)
                    input(
                        inputType = InputType.TYPE_CLASS_NUMBER,
                        prefill = getIntPreference("min_pin_length", 4).toString()
                    ) { _, charSequence ->
                        setPreference("min_pin_length", Integer.parseInt(charSequence.toString()))
                    }
                }
                return true
            }
            "lock_at_boot" -> {
                showPasswordDialog(
                    LockAction.LockType.valueOf(
                        getStringPreference(
                            "lock_at_boot",
                            LockAction.LockType.UNCHANGED.name
                        )
                    )
                ) { which ->
                    onLockTypeSelected(positionToLockType(which))
                }
            }
        }
        return super.onPreferenceTreeClick(preference)
    }

    private fun onLockTypeSelected(lockType: LockAction.LockType) {
        if (lockType == LockAction.LockType.UNCHANGED) {
            setPreference("lock_at_boot", LockAction.LockType.UNCHANGED.name)
            updateSummary(LockAction.LockType.UNCHANGED)
        } else {
            checkAdminPermission(
                onGranted = {
                    when (lockType) {
                        LockAction.LockType.PIN -> showPinChooser(REQUEST_CODE_PIN)
                        LockAction.LockType.PASSWORD -> showPasswordChooser(REQUEST_CODE_PASSWORD)
                        LockAction.LockType.SWIPE -> {
                            setPreference("lock_at_boot", LockAction.LockType.SWIPE.name)
                            updateSummary(LockAction.LockType.SWIPE)
                        }
                        else -> throw IllegalStateException("Unknown lock type $lockType.")
                    }
                },
                onRationaleNeeded = {
                    lockTypeIfGranted = lockType
                    showAdminPermissionRationale(
                        onOk = { askAdminPermission(REQUEST_CODE_ADMIN_PERMISSION) },
                        onCancel = { onAdminPermissionDenied() }
                    )
                },
                onDenied = {
                    lockTypeIfGranted = lockType
                    askAdminPermission(REQUEST_CODE_ADMIN_PERMISSION)
                }
            )
        }
    }

    private fun updateSummary(lockType: LockAction.LockType) {
        findPreference<Preference>("lock_at_boot")?.setSummary(lockTypeToTextRes(lockType))
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REQUEST_CODE_PIN -> if (resultCode == Activity.RESULT_OK) {
                AppPreferencesHelper.lockAtBoot = LockAction.LockType.PIN
                AppPreferencesHelper.lockAtBootInput = data!!.getStringExtra("input")!!
                updateSummary(LockAction.LockType.PIN)
            }
            REQUEST_CODE_PASSWORD -> if (resultCode == Activity.RESULT_OK) {
                AppPreferencesHelper.lockAtBoot = LockAction.LockType.PASSWORD
                AppPreferencesHelper.lockAtBootInput = data!!.getStringExtra("input")!!
                updateSummary(LockAction.LockType.PASSWORD)
            }
            REQUEST_CODE_ADMIN_PERMISSION -> handleAdminPermissionResult(
                resultCode = resultCode,
                onGranted = {
                    Toast.makeText(
                        context,
                        R.string.admin_permission_granted_msg,
                        Toast.LENGTH_SHORT
                    ).show()
                    when (lockTypeIfGranted) {
                        LockAction.LockType.PIN -> showPinChooser(REQUEST_CODE_PASSWORD)
                        LockAction.LockType.PASSWORD -> showPasswordChooser(REQUEST_CODE_PIN)
                        else -> {
                        } // do nothing
                    }
                },
                onDenied = { onAdminPermissionDenied() })
        }
    }

    private fun setPreference(key: String, value: String, updateSummary: Boolean = false) {
        preferenceManager.sharedPreferences.edit().putString(key, value).apply()
        if (updateSummary)
            findPreference<Preference>(key)?.summary = value
    }

    private fun setPreference(key: String, value: Int, updateSummary: Boolean = false) {
        preferenceManager.sharedPreferences.edit().putInt(key, value).apply()
        if (updateSummary)
            findPreference<Preference>(key)?.summary = Integer.toString(value)
    }

    private fun getStringPreference(key: String, defaultValue: String?): String {
        return preferenceManager.sharedPreferences.getString(key, defaultValue)!!
    }

    private fun getIntPreference(key: String, defaultValue: Int): Int {
        return preferenceManager.sharedPreferences.getInt(key, defaultValue)
    }

    private fun onAdminPermissionDenied() {
        Toast.makeText(context, "Admin permission denied", Toast.LENGTH_SHORT).show()
    }

    companion object {
        private const val REQUEST_CODE_PIN = 1
        private const val REQUEST_CODE_PASSWORD = 2
        private const val REQUEST_CODE_ADMIN_PERMISSION = 3
    }
}
